symbol.c

```diff
@@ -316,31 +316,35 @@ rb_enc_symname_type(const char *name, long len, rb_encoding *enc, unsigned int a
 
       default:
 	type = rb_sym_constant_char_p(m, e-m, enc) ? ID_CONST : ID_LOCAL;
-      id:
-	if (m >= e || (*m != '_' && !ISALPHA(*m) && ISASCII(*m))) {
-	    if (len > 1 && *(e-1) == '=') {
-		type = rb_enc_symname_type(name, len-1, enc, allowed_attrset);
-		if (type != ID_ATTRSET) return ID_ATTRSET;
-	    }
-	    return -1;
-	}
-	while (m < e && is_identchar(m, e, enc)) m += rb_enc_mbclen(m, e, enc);
-	if (m >= e) break;
-	switch (*m) {
-	  case '!': case '?':
-	    if (type == ID_GLOBAL || type == ID_CLASS || type == ID_INSTANCE) return -1;
-	    type = ID_JUNK;
-	    ++m;
-	    if (m + 1 < e || *m != '=') break;
-	    /* fall through */
-	  case '=':
-	    if (!(allowed_attrset & (1U << type))) return -1;
-	    type = ID_ATTRSET;
-	    ++m;
-	    break;
-	}
-	break;
+        goto id;
+    }
+    goto stophere;
```

Comment:
Is foxtrot like this really better?

Reply:
I agree this is something not very clean.

Reply:
https://github.com/ruby/ruby/pull/3247/commits/12a56d74cad9ae2158a1485023ecee590ac67a73